### PR TITLE
Improve Write Failure Message

### PR DESCRIPTION
**Pull request by tgroh**
Be sure to do all of the following to help us incorporate your contribution
quickly and easily:

 - [x] Make sure the PR title is formatted like:
   `[BEAM-<Jira issue #>] Description of pull request`
 - [x] Make sure tests pass via `mvn clean verify`. (Even better, enable
       Travis-CI on your fork and ensure the whole test matrix passes).
 - [x] Replace `<Jira issue #>` in the title with the actual Jira issue
       number, if there is one.
 - [x] If this contribution is large, please file an Apache
       [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.txt).

---

If provided with an Unbounded PCollection, Write will fail due to
restriction of calling finalize only once. This error message fails in a
deep stack trace based on it not being possible to apply a GroupByKey.
Instead, throw immediately on application with a specific error message.

### CODE REVIEW

**Comment by tgroh** [comment]
R: @dhalperi

**Comment by lukecwik** [comment]
R: @lukecwik
LGTM

File: sdks/java/core/src/test/java/org/apache/beam/sdk/io/WriteTest.java [hunk 1]
```diff
@@ -289,6 +292,19 @@ public void populateDisplayData(DisplayData.Builder builder) {
     assertThat(displayData, hasDisplayItem("numShards", 1));
   }
 
+  @Test
+  public void testWriteUnbounded() {
```
Comment (lukecwik) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/io/WriteTest.java]: Expected: (an instance of java.lang.IllegalArgumentException and exception with message a string containing "Write can only be applied to a Bounded PCollection")
     but: an instance of java.lang.IllegalArgumentException <java.lang.IllegalStateException: Producer unknown for input: <unnamed> [PCollection]> is a java.lang.IllegalStateException
Stacktrace was: java.lang.IllegalStateException: Producer unknown for input: <unnamed> [PCollection]
 at com.google.common.base.Preconditions.checkState(Preconditions.java:199)
 at org.apache.beam.sdk.runners.TransformHierarchy.addInput(TransformHierarchy.java:79)
 at org.apache.beam.sdk.Pipeline.applyInternal(Pipeline.java:395)

**Comment by tgroh** [comment]
My improper positioning of the checkArgument is repaired.

File: sdks/java/core/src/test/java/org/apache/beam/sdk/io/WriteTest.java [hunk 2]
```diff
@@ -295,9 +294,8 @@ public void populateDisplayData(DisplayData.Builder builder) {
   @Test
   public void testWriteUnbounded() {
     TestPipeline p = TestPipeline.create();
-    PCollection<String> unbounded = PCollection.<String>createPrimitiveOutputInternal(p,
-        WindowingStrategy.globalDefault(),
-        IsBounded.UNBOUNDED).setCoder(StringUtf8Coder.of());
+    PCollection<String> unbounded =
+        p.apply(Create.of("foo")).setIsBoundedInternal(IsBounded.UNBOUNDED);
```
Comment (lukecwik) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/io/WriteTest.java]: Why not use unboundedreadfromboundedsource around create?

**Comment by lukecwik** [comment]
We could get rid of the StreamingWrite override in DataflowRunner now.

File: sdks/java/core/src/test/java/org/apache/beam/sdk/io/WriteTest.java (hunk 2 above)
Comment (tgroh) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/io/WriteTest.java]: streaming as true doesn't change the output of create;

`UnboundedReadFromBoundedSource` is in runners-core (so isn't visible) and produces an `IsBounded.BOUNDED PCollection` as the output of the composite.

**Comment by tgroh** [comment]
CC: @dpmills

**Comment by lukecwik** [comment]
Test failure:
org.apache.beam.sdk.io.WriteTest.testWriteUnbounded

java.lang.IllegalArgumentException: unable to serialize org.apache.beam.sdk.transforms.DoFnReflector$SimpleDoFnAdapter@34518f0
	at org.apache.beam.sdk.util.SerializableUtils.serializeToByteArray(SerializableUtils.java:56)
	at org.apache.beam.sdk.util.SerializableUtils.clone(SerializableUtils.java:93)
	at org.apache.beam.sdk.transforms.ParDo$Bound.<init>(ParDo.java:672)
	at org.apache.beam.sdk.transforms.ParDo$Unbound.of(ParDo.java:630)

**Comment by bjchambers** [comment]
`Caused by: java.io.NotSerializableException: org.apache.beam.sdk.io.WriteTest`

The enclosing class needs to be `Serializable` if you're using anonymous `DoFn`s.

**Comment by tgroh** [comment]
Fixed serialization.

**Comment by tgroh** [comment]
PTAL

Travis failure unrelated.

**Pull request closed without merging**